expression/scalar_function.go

```diff
@@ -438,13 +439,120 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
 
 // HashCode implements Expression interface.
 func (sf *ScalarFunction) HashCode(sc *stmtctx.StatementContext) []byte {
+	if sc.CanonicalHashCode.Load() {
+		simpleCanonicalizedHashCode(sf, sc)
+		return sf.hashcode
+	}
 	if len(sf.hashcode) > 0 {
 		return sf.hashcode
 	}
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+// ExpressionsSemanticEqual is used to judge whether two expression tree is semantic equivalent.
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
+	sc := ctx.GetSessionVars().StmtCtx
+	sc.CanonicalHashCode.Store(true)
+	defer sc.CanonicalHashCode.Store(false)
+	return bytes.Equal(expr1.HashCode(sc), expr2.HashCode(sc))
+}
+
+// canonicalizedHashCode is used to judge whether two expression is semantically equal.
+func simpleCanonicalizedHashCode(sf *ScalarFunction, sc *stmtctx.StatementContext) {
+	sf.hashcode = sf.hashcode[:0]
+	sf.hashcode = append(sf.hashcode, scalarFunctionFlag)
+
+	argsHashCode := make([][]byte, 0, len(sf.GetArgs()))
+	for _, arg := range sf.GetArgs() {
+		argsHashCode = append(argsHashCode, arg.HashCode(sc))
+	}
+	switch sf.FuncName.L {
+	case ast.Plus, ast.Mul, ast.EQ, ast.In, ast.LogicOr, ast.LogicAnd:
+		// encode original function name.
+		sf.hashcode = codec.EncodeCompactBytes(sf.hashcode, hack.Slice(sf.FuncName.L))
+		// reorder parameters hashcode, eg: a+b and b+a should has the same hashcode here.
+		sort.Slice(argsHashCode, func(i, j int) bool {
+			return bytes.Compare(argsHashCode[i], argsHashCode[j]) <= 0
+		})
+		for _, argCode := range argsHashCode {
+			sf.hashcode = append(sf.hashcode, argCode...)
+		}
+
+	case ast.GE, ast.LE: // directed binary OP: a >= b and b <= a should have the same hashcode.
+		// encode GE function name.
+		sf.hashcode = codec.EncodeCompactBytes(sf.hashcode, hack.Slice(ast.GE))
+		// encode GE function name and switch the args order.
+		if sf.FuncName.L == ast.GE {
+			for _, argCode := range argsHashCode {
+				sf.hashcode = append(sf.hashcode, argCode...)
+			}
+		} else {
+			for i := len(argsHashCode) - 1; i >= 0; i-- {
+				sf.hashcode = append(sf.hashcode, argsHashCode[i]...)
+			}
+		}
+	case ast.GT, ast.LT:
+		sf.hashcode = codec.EncodeCompactBytes(sf.hashcode, hack.Slice(ast.GT))
+		if sf.FuncName.L == ast.GT {
+			for _, argCode := range argsHashCode {
+				sf.hashcode = append(sf.hashcode, argCode...)
+			}
+		} else {
+			for i := len(argsHashCode) - 1; i >= 0; i-- {
+				sf.hashcode = append(sf.hashcode, argsHashCode[i]...)
+			}
+		}
+	case ast.UnaryNot:
+		child, ok := sf.GetArgs()[0].(*ScalarFunction)
+		if !ok {
+			// use the origin arg hash code.
+			for _, argCode := range argsHashCode {
+				sf.hashcode = append(sf.hashcode, argCode...)
```

Comment:
Do we need to encode the not op?

Reply:
Nope：
Expression1: not(a<b) 
Expression2: a>=b 
Expression1 and Expression2 should have the same hashcode, we try to disintegrate the NOT into a mathematical operation convertion.